devito/function.py

```diff
@@ -815,6 +786,68 @@ def __indices_setup__(cls, **kwargs):
     def __shape_setup__(cls, **kwargs):
         return kwargs.get('shape', (kwargs.get('npoint'),))
 
+    def _arg_defaults(self, alias=None):
+        """
+        Returns a map of default argument values defined by this symbol.
+
+        :param alias: (Optional) name under which to store values.
+        """
+        key = alias or self
+        args = super(AbstractSparseFunction, self)._arg_defaults(alias=alias)
+        for child_name in self._child_functions:
+            child = getattr(self, child_name)
+            args.update(child._arg_defaults(alias=getattr(key, child_name)))
+        return args
+
+    @property
+    def _arg_names(self):
+        """Return a tuple of argument names introduced by this function."""
+        return tuple([self.name] + [x for x in self._child_functions])
+
+
+class SparseFunction(AbstractSparseFunction):
+    """
+    A special :class:`TensorFunction` representing a set of sparse point
+    objects that are not aligned with the computational grid.
+
+    A :class:`SparseFunction` provides symbolic interpolation routines
+    to convert between grid-aligned :class:`Function` objects and sparse
+    data points.
+
+    :param name: Name of the function.
+    :param npoint: Number of points to sample.
+    :param grid: :class:`Grid` object defining the computational domain.
+    :param shape: (Optional) shape of the function. Defaults to ``(npoints,)``.
+    :param dimensions: (Optional) symbolic dimensions that define the
+                       data layout and function indices of this symbol.
+    :param coordinates: (Optional) coordinate data for the sparse points.
+    :param space_order: Discretisation order for space derivatives.
+    :param dtype: Data type of the buffered data.
+    :param initializer: (Optional) A callable to initialize the data
+
+    .. note::
+
+        The parameters must always be given as keyword arguments, since
+        SymPy uses `*args` to (re-)create the dimension arguments of the
+        symbolic function.
+    """
+
+    is_SparseFunction = True
+    _child_functions = ['coordinates']
+
+    def __init__(self, *args, **kwargs):
+        if not self._cached():
+            super(SparseFunction, self).__init__(*args, **kwargs)
+
+            # Set up coordinates of sparse points
+            coordinates = Function(name='%s_coords' % self.name,
+                                   dimensions=(self.indices[-1], Dimension(name='d')),
+                                   shape=(self.npoint, self.grid.dim), space_order=0)
+            coordinate_data = kwargs.get('coordinates')
+            if coordinate_data is not None:
+                coordinates.data[:] = coordinate_data[:]
```

Comment:
random question unrelated to this PR: do we really want to copy data, or would we rather want a ``SparseFunction`` to take ownership of what it's been given?

Reply:
I don't follow what you mean. Care to elaborate?